### PR TITLE
inline single element h2d

### PR DESCRIPTION
**Pull request by Yancey1989**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_lower_to_library_call.cc [hunk 1]
```diff
@@ -104,6 +104,12 @@ void InsertSyncOnStream(Operation* op, Value ctx, Value stream_handle,
                               "sync_on_stream", false, "gpu");
 }
 
+// return true if op in fusion block
+bool isInFusionBlock(Operation* op) {
```
Comment (tanyokwok) [inline tao_compiler/mlir/disc/transforms/disc_lower_to_library_call.cc:108]: Nit: move to `fusion_utils.cc`

File: tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc [hunk 2]
```diff
@@ -959,6 +960,16 @@ Value elementalLower<lmhlo::CopyOp>(OpBuilder* b, Location loc,
   return result;
 }
 
+template <>
+Value elementalLower<lmhlo_disc::H2DOp>(OpBuilder* b, Location loc,
```
Comment (tanyokwok) [inline tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc:964]: Nit: Add some unit tests

File: tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc [hunk 3]
```diff
@@ -959,6 +960,16 @@ Value elementalLower<lmhlo::CopyOp>(OpBuilder* b, Location loc,
...
+                                        lmhlo_disc::H2DOp op,
+                                        ValueRange output_index,
+                                        bool check_cache,
+                                        LowerConfig* lower_config) {
+  Value zero = b->create<arith::ConstantIndexOp>(loc, 0);
+  return b->create<memref::LoadOp>(loc, op->getOperand(0), zero);
```
Comment (tanyokwok) [inline tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc:970]: Does that mean we will use uniform memory access, and direct access to CPU data from the device?

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc [hunk 4]
```diff
@@ -1337,13 +1338,27 @@ bool BaseCpuFusionStrategy::tryFuse(ShapeAnalysis& shapeAnalysis,
 ////////////////////// Base GPU FusionStrategy Implementation /////////
 //////////////////////////////////////////////////////////////////
 
+bool isSingleElementH2DOp(Operation* op) {
+  auto h2d_op = dyn_cast<lmhlo_disc::H2DOp>(op);
+  if (!h2d_op) return false;
+
+  int rank = op->getOperand(0).getType().cast<MemRefType>().getRank();
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/fusion_utils.cc:1345]: using `getNumElements() == 1` to support other rank?

**Review by wyzero — COMMENTED** [review]
add e2e UT？

File: tao_compiler/mlir/disc/transforms/disc_lower_to_library_call.cc (hunk 1 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/disc_lower_to_library_call.cc:108]: this function is only used in `disc-lower-to-library-call-pass`, maybe here is okay, let's move to fusion_utils.cc if it's a common function.

File: tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc (hunk 2 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc:964]: done

File: tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc (hunk 3 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc:970]: yes, `gpu-kernel-outlining` will move single element CPU buffer to kernel parameter.

**Comment by Yancey1989** [comment]
> add e2e UT as well？

@wyzero thanks, added a e2e ut.

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc (hunk 4 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/fusion_utils.cc]: done

**Pull request merged**